rustworkx-core/src/shortest_path/all_shortest_paths.rs

```diff
@@ -0,0 +1,207 @@
+use petgraph::Direction::Incoming;
+
+use super::dijkstra;
+use crate::dictmap::*;
+
+/// Dijkstra-based all shortest paths algorithm.
+///
+/// Compute every single shortest path from `start` to `goal`.
+///
+/// The graph should be [`Visitable`] and implement [`IntoEdges`]. The function
+/// `edge_cost` should return the cost for a particular edge, which is used
+/// to compute path costs. Edge costs must be non-negative.
+///
+///
+/// Returns a [`HashSet`] which contains all possible shortest paths. Each path
+/// is a Vec of node indices of the path starting with `start` and ending `goal`.
+/// # Example
+/// ```rust
+/// use rustworkx_core::petgraph::Graph;
+/// use rustworkx_core::petgraph::prelude::*;
+/// use rustworkx_core::dictmap::DictMap;
+/// use rustworkx_core::shortest_path::all_shortest_paths;
+/// use rustworkx_core::Result;
+/// use ahash::HashSet;
+///
+/// let mut graph : Graph<(), (), Directed>= Graph::new();
+/// let a = graph.add_node(()); // node with no weight
+/// let b = graph.add_node(());
+/// let c = graph.add_node(());
+/// let d = graph.add_node(());
+/// let e = graph.add_node(());
+/// let f = graph.add_node(());
+/// let g = graph.add_node(());
+/// // z will be in another connected component
+/// let z = graph.add_node(());
+///
+/// graph.extend_with_edges(&[
+///     (a, b),
+///     (a, c),
+///     (b, d),
+///     (b, f),
+///     (c, d),
+///     (d, e),
+///     (f, e),
+///     (e, g)    
+/// ]);
+/// // a ----> b ----> f
+/// // |       |       |
+/// // v       v       v
+/// // c ----> d ----> e ----> g
+///
+/// let expected_res: HashSet<Vec<NodeIndex>>= [
+///      vec![a, b, f, e, g],
+///      vec![a, b, d, e, g],
+///      vec![a, c, d, e, g]
+///     ].into_iter().collect();
+/// let res: Result<HashSet<Vec<NodeIndex>>> = all_shortest_paths(
+///     &graph, a, g, |_| Ok(1)
+/// );
+/// assert_eq!(res.unwrap(), expected_res)
+/// ```
+pub fn all_shortest_paths<G, F, E, K>(
+    graph: G,
+    start: G::NodeId,
+    goal: G::NodeId,
+    mut edge_cost: F,
+) -> Result<HashSet<Vec<G::NodeId>>, E>
+where
+    G: IntoEdgesDirected + Visitable + NodeIndexable,
+    G::NodeId: Eq + Hash,
+    F: FnMut(G::EdgeRef) -> Result<K, E>,
+    K: Measure + Copy,
+{
+    let scores: DictMap<G::NodeId, K> = dijkstra(&graph, start, Some(goal), &mut edge_cost, None)?;
+    if !scores.contains_key(&goal) {
+        return Ok(HashSet::default());
+    }
+    let mut paths = HashSet::new();
+    let path = VecDeque::from([goal]);
+    let mut queue = vec![(goal, path)];
+    while let Some((curr, curr_path)) = queue.pop() {
+        let curr_dist = *scores.get(&curr).unwrap();
+        for edge in graph.edges_directed(curr, Incoming) {
+            let next_dist = match scores.get(&edge.source()) {
+                Some(x) => *x,
+                None => continue,
+            };
+            if curr_dist == next_dist + edge_cost(edge)? {
```

Comment:
This is fine but I wonder if just generating the graph with these edges + calling `all_simple_paths` that we already implemented does the job

Reply:
I think this is possible, but since we need to backtrack anyway, we're already doing most of the work needed to generate the paths. Calling `all_simple_paths` would mean enumerating over all shortest path edges once more. But if you prefer to return an `Impl Iterator<>` I can definitely change this.